File: crates/timekeep-core/src/model/audit.rs
//! Audit log domain model.
//!
//! Every authenticated API request is captured automatically by
//! middleware — no handler code changes needed. The audit log
//! provides a complete, immutable timeline of system activity.
//!
//! ## Design
//!
//! ```text
//! Request → JWT middleware → Audit middleware → Handler → Response
//!                                │
//!                                └── record_audit(AuditEvent {
//!                                      actor, action, resource, status, ...
//!                                    })
//! ```
//!
//! ## Actions
//!
//! Action strings follow the pattern `{resource}.{operation}`:
//! - `device.created`, `device.updated`, `device.deleted`
//! - `punch.corrected`
//! - `endpoint.created`, `endpoint.updated`, `endpoint.deleted`
//! - `settings.updated`
//! - `auth.login`, `auth.login_failed`

use serde::{Deserialize, Serialize};

/// A single audit log entry — who did what, when, and the result.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuditEvent {
    /// Unique identifier (UUID v7).
    pub id: String,

    /// Unix timestamp (seconds).
    pub timestamp: i64,

    /// Username of who performed the action ("system" for automated actions).
    pub actor: String,

    /// What happened: "device.created", "punch.corrected", "settings.updated", etc.
    pub action: String,

    /// What was affected: "device/SN001", "punch/abc123", "settings".
    pub resource: String,

    /// Action-specific detail as JSON (e.g., device label, old/new values).
    pub detail: Option<serde_json::Value>,

    /// IP address of the client that made the request.
    pub ip_address: Option<String>,

    /// "success" or "error".
    pub status: String,

    /// Error message if status is "error".
    pub error_message: Option<String>,
}

impl AuditEvent {
    /// Create a new audit event with the current timestamp.
    pub fn new(actor: String, action: String, resource: String, status: String) -> Self {
        Self {
            id: uuid::Uuid::now_v7().to_string(),
            timestamp: jiff::Timestamp::now().as_second(),
            actor,
            action,
            resource,
            detail: None,
            ip_address: None,
            status,
            error_message: None,
        }
    }

    /// Builder: attach request detail.
    pub fn with_detail(mut self, detail: serde_json::Value) -> Self {
        self.detail = Some(detail);
        self
    }

    /// Builder: attach IP address.
    pub fn with_ip(mut self, ip: String) -> Self {
        self.ip_address = Some(ip);
        self
    }

    /// Builder: attach error message.
    pub fn with_error(mut self, error: String) -> Self {
        self.error_message = Some(error);
        self
    }
}

/// Filters for querying audit logs.
#[derive(Debug, Clone, Default)]
pub struct AuditFilter {
    /// Filter by actor username.
    pub actor: Option<String>,

    /// Filter by action prefix (e.g., "device." matches all device actions).
    pub action: Option<String>,

    /// Filter by resource path.
    pub resource: Option<String>,

    /// Only events after this timestamp (inclusive).
    pub since: Option<jiff::Timestamp>,

    /// Only events before this timestamp (inclusive).
    pub until: Option<jiff::Timestamp>,

    /// Search across actor, action, and resource.
    pub search: Option<String>,

    /// Sort field: "timestamp" (default).
    pub sort_by: String,

    /// Sort direction.
    pub sort_order: crate::query::SortOrder,

    /// Items per page. Default: 50.
    pub limit: u32,

    /// Cursor for pagination.
    pub cursor: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_audit_event_new() {
        let event = AuditEvent::new(
            "admin".into(),
            "device.created".into(),
            "device/SN001".into(),
            "success".into(),
        );
        assert_eq!(event.actor, "admin");
        assert_eq!(event.action, "device.created");
        assert_eq!(event.status, "success");
        assert!(!event.id.is_empty());
        assert!(event.timestamp > 0);
    }

    #[test]
    fn test_audit_event_builders() {
        let event =
            AuditEvent::new("admin".into(), "test".into(), "test/1".into(), "success".into())
                .with_ip("192.168.1.1".into())
                .with_detail(serde_json::json!({"key": "value"}))
                .with_error("something went wrong".into());

        assert_eq!(event.ip_address, Some("192.168.1.1".into()));
        assert!(event.detail.is_some());
        assert_eq!(event.error_message, Some("something went wrong".into()));
    }
}

File: dashboard/src/lib/api/audit.ts
import { apiGet, toUnixSeconds } from "./client";
import type { FacetGroup } from "./client";
import type { EntitySchema } from "@/types/metadata";

// ── Types ──────────────────────────────────────────────────────────────────

/** Matches the Rust `AuditEventResponse` DTO. */
export type AuditEvent = {
  id: string;
  timestamp: number;
  actor: string;
  action: string;
  resource: string;
  detail?: Record<string, unknown> | null;
  ip_address?: string | null;
  status: string;
  error_message?: string | null;
};

export type AuditFilter = {
  actor?: string;
  action?: string;
  resource?: string;
  since?: string;
  until?: string;
  search?: string;
  sort_by?: string;
  sort_order?: "asc" | "desc";
  limit?: number;
  cursor?: string;
};

function buildAuditParams(filter: AuditFilter): string {
  const params = new URLSearchParams();
  if (filter.actor) params.set("actor", filter.actor);
  if (filter.action) params.set("action", filter.action);
  if (filter.resource) params.set("resource", filter.resource);
  if (filter.since) params.set("since", toUnixSeconds(filter.since));
  if (filter.until) params.set("until", toUnixSeconds(filter.until));
  if (filter.search) params.set("search", filter.search);
  if (filter.sort_by) params.set("sort_by", filter.sort_by);
  if (filter.sort_order) params.set("sort_order", filter.sort_order);
  if (filter.limit !== undefined) params.set("limit", String(filter.limit));
  if (filter.cursor) params.set("cursor", filter.cursor);
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

/** Query audit logs. Requires Viewer+. */
export function fetchAuditLogs(filter: AuditFilter = {}): Promise<AuditEvent[]> {
  return apiGet<AuditEvent[]>(`audit${buildAuditParams(filter)}`).json();
}

/** Fetch a single audit event by ID. Requires Viewer+. */
export function fetchAuditEvent(id: string): Promise<AuditEvent> {
  return apiGet<AuditEvent>(`audit/${encodeURIComponent(id)}`).json();
}

// ── Schema (Metadata System) ────────────────────────────────────────────────

/** Fetch entity schema for audit logs (column metadata, sortability, filterability). */
export function fetchAuditSchema(): Promise<EntitySchema> {
  return apiGet<EntitySchema>("audit/schema").json();
}

/**
 * Facet filter params for audit queries.
 *
 * Matches the Rust facet endpoint at GET /api/audit/filters.
 */
export type AuditFacetParams = {
  dimension?: string;
  search?: string;
  since?: string;
  until?: string;
  limit?: number;
};

function buildAuditFacetParams(filter: AuditFacetParams): string {
  const params = new URLSearchParams();
  if (filter.dimension) params.set("dimension", filter.dimension);
  if (filter.search) params.set("search", filter.search);
  if (filter.since) params.set("since", toUnixSeconds(filter.since));
  if (filter.until) params.set("until", toUnixSeconds(filter.until));
  if (filter.limit !== undefined) params.set("limit", String(filter.limit));
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

/** Fetch faceted filter metadata for audit queries. */
export function fetchAuditFilters(filter: AuditFacetParams = {}): Promise<FacetGroup[]> {
  return apiGet<FacetGroup[]>(`audit/filters${buildAuditFacetParams(filter)}`).json();
}
